### PR TITLE
Support for Views, Materialized Views, Foreign Tables

### LINKED ISSUE #205: Allow views to be queried via GQL
# Summary
[summary]: #summary

Make views available in the GQL schema so they can be queried.

# Rationale
[rationale]: #rationale

I have a view with some complex filtering not available via the GQL filter. I would like to be able to query this via GQL like a normal table rather than having to use supabase js.

# Drawbacks
[drawbacks]: #drawbacks

#69 I noticed that views were mentioned as being explicitly excluded in this issue so I presume that there are some drawbacks, but my SQL experience isn't enough to be aware of what these may be?

### PR DESCRIPTION
**Pull request by olirice**
## What kind of change does this PR introduce?
Adds support for 
- Views
- Materialized Views
- Foreign Tables

with the requirement that they specify a comment directive with the intended primary key
```sql
comment on view person is e'@graphql({"primary_key_columns": ["id"]})';
```

Since no foreign keys are present, the types for view entities are not connected to the rest of the graph. Relationships can be defined comment directives.

```sql
comment on view "EmailAddress" is e'
    @graphql({
        "foreign_keys": [
          {
            "local_name": "addresses",
            "local_columns": ["accountId"],
            "foreign_name": "account",
            "foreign_schema": "public",
            "foreign_table": "Account",
            "foreign_columns": ["id"]
          }
        ]
    })';
```

resolves #205

### CODE REVIEW

File: sql/load_sql_context.sql [hunk 1]
```diff
@@ -345,7 +304,12 @@ select
                             pg_class pc
                         where
                             pc.relnamespace = pn.oid
-                            and pc.relkind = 'r' -- tables
+                            and pc.relkind in (
+                                'r', -- table
+                                'v', -- view
+                                'm', -- mat view
+                                'f'  -- foreign table
+                            )
```
Comment (olirice) [inline sql/load_sql_context.sql:312]: enabling more types of entities

File: src/sql_types.rs [hunk 2]
```diff
@@ -89,104 +90,125 @@ pub struct Enum {
...
     pub indexes: Vec<Index>,
     pub functions: Vec<Arc<Function>>,
-    pub foreign_keys: Vec<Arc<ForeignKey>>,
     pub directives: TableDirectives,
 }
 
 impl Table {
-    pub fn primary_key(&self) -> Option<&Index> {
-        self.indexes.iter().find(|x| x.is_primary_key)
+    pub fn primary_key(&self) -> Option<Index> {
```
Comment (olirice) [inline src/sql_types.rs:166]: extended to out primary key defined in a comment directive if no native primary key is found

**Review by alaister — APPROVED** [review]

**Comment by drewbietron** [comment]
@olirice - Any word when a release may be cut that includes this?  

Also just I understand correctly, the new release that needs to be cut is a new Supabase CLI release if I want to be able to use this locally?

🙏

**Comment by olirice** [comment]
The next pg_graphql release will be this week. There's a bugfix that I expect to merge NLT Wednesday and then I'll push to have a release cut.

At that point it gets built into the images produced by supabase/postgres. I'll mention in that PR that the update is needed on dockerhub for the CLI. Est 2-3 weeks from now before it out fully

**Comment by drewbietron** [comment]
@olirice - Saw that the 1.1 release was merged.  Is it possible to use it locally and in production without a CLI update?

**Comment by olirice** [comment]
@drewbietron the versions hosted on the supabase platform are not connected to your locally installed CLI version

- For new projects, `1.1.0` is installed by default
- For existing projects, it is available, but you have to run 
```sql
drop extension pg_graphql; create extension pg_graphql;
```
to update receive the update
- to use it in the CLI you would have to update the CLI to the most recent version (the docker image is pinned in src)

**Comment by drewbietron** [comment]
@olirice - Updated the CLI to the latest (`1.36.7`) but it installs pg_graphql v 0.5.0 if I query ```SELECT * 
FROM pg_extension;``` in local studio. I don't see a version of pg_graphql anywhere in the CLI repo, but I just grepped for pg_graphl so it may be somewhere else?  Sorry for all of the questions, but it looks like this should all be out and working but Im not able to get anything to happen.

Here's an example of the schema I'm testing this on, which I'm not seeing any types for shipping policy profiles created?
```sql
CREATE TABLE shipping_policies (
  id uuid DEFAULT uuid_generate_v4 () PRIMARY KEY,
  created_at timestamp DEFAULT NOW() NOT NULL,
  updated_at timestamp DEFAULT NOW() NOT NULL,
  name text,
  free_shipping boolean DEFAULT FALSE NOT NULL,
  account_covered_price bigint CHECK (account_covered_price >= 1500),
  shipping_carriers shipping_carriers ARRAY NOT NULL,
  package_height real NOT NULL,
  package_width real NOT NULL,
  package_length real NOT NULL,
  package_dimension_unit shipping_package_dimension_units NOT NULL,
  package_weight real NOT NULL,
  package_weight_unit shipping_package_weight_units NOT NULL,
  account_id uuid NOT NULL,
  address_id uuid NOT NULL
);

CREATE VIEW shipping_policy_profiles AS
SELECT
  id,
  free_shipping,
  account_covered_price,
  shipping_carriers
FROM
  public.shipping_policies;

COMMENT ON VIEW shipping_policy_profiles IS e'@graphql({"primary_key_columns": ["id"]})';
```

**Comment by olirice** [comment]
@drewbietron when you run supabase locally via `supabase start` and then run `docker ps` what version of `supabase/postgres` do you see running? The latest version with pg_graphql `1.1.0` is `supabase/postgres:15.1.0.33`

If you see a version lower than that, let me know, and I'll open an issue on `supabase/cli` asking for the recommended approach for updating an existing a local project with the latest docker image that you can track against

**Comment by drewbietron** [comment]
> @drewbietron when you run supabase locally via `supabase start` and then run `docker ps` what version of `supabase/postgres` do you see running? The latest version with pg_graphql `1.1.0` is `supabase/postgres:15.1.0.33`
> 
> If you see a version lower than that, let me know, and I'll open an issue on `supabase/cli` asking for the recommended approach for updating an existing a local project with the latest docker image that you can track against

The CLI was pinned to an older version of supabase/postgres.  I posted an issue in that repo asking to have it bumped - https://github.com/supabase/cli/issues/848 -- I ended up just running the CLI locally and declaring the latest version of supabase/postgres which allows me to use this PR.  Sort of a hacky work around but understand how it can be complex to sync up all of the different dependent repos.  Hoping that there can be some options to explicitly declare a version of all dependent packages in the config.toml or something equivalent so that local dev or self hosting can be a little smoother.

Have been working in views throughout the day today and its great.  Really appreciative that you were able to get this in!

**Comment by olirice** [comment]
> Have been working in views throughout the day today and its great

🙌

> running the CLI locally and declaring the latest version of supabase/postgres

does that mean you built the CLI from source?

>  I ended up just running the CLI locally and declaring the latest version of supabase/postgres which allows me to use this PR

I just started a new project with cli version 1.37.1 and it came with `supabase/postgres:15.1.0.33` which included `pg_graphql=1.1.0` so I think that part is working correctly. Was your project created on an older image version, or was it new?

in any case, glad you got it working!

**Comment by drewbietron** [comment]
> does that mean you built the CLI from source?
> 
> Yeah, just changed the pg15 image version and it all seems to work fine

**Comment by HTMHell** [comment]
> > does that mean you built the CLI from source?
> 
> Yeah, just changed the pg15 image version and it all seems to work fine

Can you elaborate on how one can do that?
@olirice I have an old project with `postgres:14.1.0.89`, and updating the CLI does not upgrade it, so I can't run it locally.  Can you advise?

**Comment by olirice** [comment]
Once the main platform goes on to the latest major version of postgres (15), previous versions  no longer receive extension upgrades except in cases of security vulnerabilities. Given that, you'd have to [upgrade your project](https://supabase.com/docs/guides/platform/migrating-and-upgrading-projects) to get the latest pg_graphql version

**Pull request merged**